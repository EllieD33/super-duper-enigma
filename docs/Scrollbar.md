# Scrollbar Component

## Description

The `Scrollbar` component provides a custom scrollbar for horizontally scrollable content. It allows users to interact with the scrollbar, including dragging the thumb and clicking on the track to scroll. The component also adjusts dynamically based on the content's size, showing or hiding the scrollbar as needed.

## Props

### `ScrollbarProps`

| Prop       | Type              | Required | Description                                                                                                       |
| ---------- | ----------------- | -------- | ----------------------------------------------------------------------------------------------------------------- |
| `children` | `React.ReactNode` | âœ…       | The content to be displayed inside the scrollable container. It can include any valid JSX elements or components. |

## Behaviour

- The scrollbar is only visible when the content overflows horizontally.
- The thumb's width and position are dynamically calculated based on the content's width and scroll position.
- Users can drag the scrollbar thumb to scroll the content, or click directly on the track to jump to a specific scroll position.
- The component automatically adjusts when the content or window size changes, ensuring the scrollbar and thumb update accordingly.

[Explore Scrollbar in Storybook](hhttp://localhost:6006/?path=/story/library-scrollbar-scrollbar--default&globals=viewport:largeTablet)

## Notes

- The component makes use of a custom scrollbar design rather than the native browser scrollbar.
- The scrollbar will only be visible if the content overflows horizontally.
- The component uses `requestAnimationFrame` to efficiently handle resizing and scrolling events.
