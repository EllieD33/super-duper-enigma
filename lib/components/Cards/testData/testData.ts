import { convertNumberToWords } from "../../../utils/convertNumberToWords";
import mockImages from "../../Galleries/MasonryGallery/testData/testData";
import { ProductCardProps } from "../ProductCard/ProductCard";

const tagOptions = [
    ["Electronics", "Gadget", "New"],
    ["Fashion", "Trending", "Sale"],
    ["Home", "Decor", "Popular"],
    ["Sports", "Outdoor", "Gear"],
    ["Books", "Education", "Read"],
    ["Food", "Gourmet", "Fresh"],
];

export const mockProducts: ProductCardProps[] = Array.from(
    { length: mockImages.length },
    (_, index) => ({
        productInfo: {
            productId: index + 1,
            productImage: mockImages[index],
            productName: `Product ${convertNumberToWords(index + 1)}`,
            productDescription:
                "This is an example product description. It's just a couple of lines of text.",
            productPrice: index + 1 * 8,
            productTags: tagOptions[index % tagOptions.length],
        },
        buttonText: "Add to basket",
    })
);
