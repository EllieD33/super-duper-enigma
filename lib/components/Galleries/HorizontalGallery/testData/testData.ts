const imageData = [
    {
        imageId: 1,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/08/08/15/01/flower-8177578_640.jpg",
    },
    {
        imageId: 2,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/04/21/11/51/flower-7941764_640.jpg",
    },
    {
        imageId: 3,
        imageUrl:
            "https://cdn.pixabay.com/photo/2024/01/17/17/22/bee-8515123_960_720.jpg",
    },
    {
        imageId: 4,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/04/11/22/08/flower-7918323_960_720.jpg",
    },
    {
        imageId: 5,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/06/04/12/00/rose-8039502_960_720.jpg",
    },
];

const mockImages = imageData.map((data) => ({
    ...data,
    altText: `image${data.imageId} alt text`,
}));

export default mockImages;
