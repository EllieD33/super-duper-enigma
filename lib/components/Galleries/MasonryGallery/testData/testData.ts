const imageData = [
    {
        imageId: 1,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/08/08/15/01/flower-8177578_640.jpg",
    },
    {
        imageId: 2,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/04/21/11/51/flower-7941764_640.jpg",
    },
    {
        imageId: 3,
        imageUrl:
            "https://cdn.pixabay.com/photo/2024/01/17/17/22/bee-8515123_960_720.jpg",
    },
    {
        imageId: 4,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/04/11/22/08/flower-7918323_960_720.jpg",
    },
    {
        imageId: 5,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/06/04/12/00/rose-8039502_960_720.jpg",
    },
    {
        imageId: 6,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/07/28/20/36/flower-8155951_960_720.jpg",
    },
    {
        imageId: 7,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/06/27/16/30/flower-8092680_960_720.jpg",
    },
    {
        imageId: 8,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/04/21/12/18/flower-7941814_960_720.jpg",
    },
    {
        imageId: 9,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/03/21/01/31/flower-7866187_960_720.jpg",
    },
    {
        imageId: 10,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/10/18/08/17/rose-8323259_960_720.jpg",
    },
    {
        imageId: 11,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/07/23/08/46/flower-8144644_960_720.jpg",
    },
    {
        imageId: 12,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/04/10/15/12/succulent-7914023_960_720.jpg",
    },
    {
        imageId: 13,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/01/09/12/49/ferns-7707348_960_720.jpg",
    },
    {
        imageId: 14,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/11/02/15/58/flower-8360946_960_720.jpg",
    },
    {
        imageId: 15,
        imageUrl:
            "https://cdn.pixabay.com/photo/2024/03/15/18/53/magnolia-flower-8635583_960_720.jpg",
    },
    {
        imageId: 16,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/11/13/01/30/flower-8384359_960_720.jpg",
    },
    {
        imageId: 17,
        imageUrl:
            "https://cdn.pixabay.com/photo/2024/03/04/16/07/winter-8612635_960_720.jpg",
    },
    {
        imageId: 18,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/01/04/19/40/grass-7697505_960_720.jpg",
    },
    {
        imageId: 19,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/10/02/05/29/flowers-8288492_960_720.jpg",
    },
    {
        imageId: 20,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/08/31/21/52/houseleek-8225994_960_720.jpg",
    },
    {
        imageId: 21,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/04/02/09/15/sempervivum-7894111_960_720.jpg",
    },
    {
        imageId: 22,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/03/19/16/15/butterfly-7862893_960_720.jpg",
    },
    {
        imageId: 23,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/09/14/22/16/fern-8253836_960_720.jpg",
    },
    {
        imageId: 24,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/07/23/06/57/leaf-8144517_960_720.jpg",
    },
    {
        imageId: 25,
        imageUrl:
            "https://cdn.pixabay.com/photo/2024/02/27/00/13/heliconia-8599119_960_720.jpg",
    },
    {
        imageId: 26,
        imageUrl:
            "https://cdn.pixabay.com/photo/2024/05/15/07/59/flowers-8763039_960_720.jpg",
    },
    {
        imageId: 27,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/07/08/19/28/leaves-8115077_960_720.jpg",
    },
    {
        imageId: 28,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/09/22/07/02/red-8268266_960_720.jpg",
    },
    {
        imageId: 29,
        imageUrl:
            "https://cdn.pixabay.com/photo/2024/02/16/14/20/snail-8577681_960_720.jpg",
    },
    {
        imageId: 30,
        imageUrl:
            "https://cdn.pixabay.com/photo/2024/01/18/16/54/leaves-8517219_960_720.jpg",
    },
    {
        imageId: 31,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/04/27/14/27/butterfly-7954767_960_720.jpg",
    },
    {
        imageId: 32,
        imageUrl:
            "https://cdn.pixabay.com/photo/2023/11/13/00/47/cactus-8384331_960_720.jpg",
    },
];

const mockImages = imageData.map((data) => ({
    ...data,
    altText: `image${data.imageId} alt text`,
}));

export default mockImages;
