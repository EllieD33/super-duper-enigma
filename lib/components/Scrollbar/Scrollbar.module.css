.scrollableContainer {
    position: relative;
    height: 100%;
    width: 100%;
    overflow: hidden;
}

.contentContainer {
    height: 100%;
    width: 100%;
    overflow-x: scroll;
    overflow-y: hidden;
    -ms-overflow-style: none;
    scrollbar-width: none;
}
.contentContainer::-webkit-scrollbar {
    display: none;
}
.scrollbar {
    width: 100%;
    height: 8px;
    background-color: gainsboro;
    border-radius: 4px;
    overflow: hidden;
    margin-top: 5px;
}

.hiddenScrollbar {
    opacity: 0;
    pointer-events: none;
}

.thumb {
    position: relative;
    height: 100%;
    background-color: cadetblue;
    border-radius: 4px;
    cursor: pointer;
    transform: translateX(0);
    transition: transform 0.1s ease;
    cursor: grab;
}
