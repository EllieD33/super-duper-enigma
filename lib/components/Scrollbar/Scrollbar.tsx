import React, {
    ReactElement,
    useRef,
    useState,
    useCallback,
    useEffect,
} from "react";
import styles from "./Scrollbar.module.css";

export interface ScrollbarProps {
    children: React.ReactNode;
}

const Scrollbar = ({ children }: ScrollbarProps): ReactElement => {
    const contentRef = useRef<HTMLDivElement>(null);
    const thumbRef = useRef<HTMLDivElement>(null);
    const trackRef = useRef<HTMLDivElement>(null);
    const thumbWidthRef = useRef<number>(20);
    const startClientXRef = useRef<number>(0);
    const startScrollLeftRef = useRef<number>(0);
    const [isScrollbarVisible, setIsScrollbarVisible] =
        useState<boolean>(false);
    const [isDragging, setIsDragging] = useState<boolean>(false);

    const handleThumbPosition = useCallback(() => {
        // istanbul ignore next
        if (!contentRef.current || !thumbRef.current || !trackRef.current)
            return;

        const { scrollLeft, scrollWidth, clientWidth } = contentRef.current;

        const trackWidth = trackRef.current.clientWidth;
        const thumbWidth = (clientWidth / scrollWidth) * trackWidth;

        thumbWidthRef.current = Math.max(thumbWidth, 20);
        thumbRef.current.style.width = `${thumbWidthRef.current}px`;
        const maxThumbLeft = trackWidth - thumbWidthRef.current;

        const thumbLeft =
            (scrollLeft / (scrollWidth - clientWidth)) * maxThumbLeft;

        thumbRef.current.style.transform = `translateX(${thumbLeft}px)`;
    }, []);

    const handleResize = useCallback(() => {
        if (!contentRef.current || !thumbRef.current || !trackRef.current)
            return;

        const { clientWidth, scrollWidth } = contentRef.current;
        const trackWidth = trackRef.current.clientWidth;

        const hasOverflow = scrollWidth > clientWidth;
        setIsScrollbarVisible(hasOverflow);

        if (hasOverflow) {
            const thumbWidth = Math.max(
                (clientWidth / scrollWidth) * trackWidth,
                20
            );
            thumbWidthRef.current = thumbWidth;
            thumbRef.current.style.width = `${thumbWidth}px`;
            handleThumbPosition();
        }
    }, [handleThumbPosition]);

    const handleTrackClick = useCallback(
        (e: React.MouseEvent<HTMLDivElement>) => {
            e.preventDefault();
            // istanbul ignore next
            if (!contentRef.current || !trackRef.current) return;

            const { left: trackLeft } =
                trackRef.current.getBoundingClientRect();
            const clickPosition = e.clientX - trackLeft;
            const { scrollWidth, clientWidth } = contentRef.current;
            const clickRatio = clickPosition / trackRef.current.clientWidth;

            contentRef.current.scrollTo({
                left: clickRatio * (scrollWidth - clientWidth),
                behavior: "smooth",
            });
        },
        []
    );

    const handleThumbMouseDown = useCallback(
        (e: React.MouseEvent<HTMLDivElement>) => {
            e.preventDefault();
            setIsDragging(true);
            startClientXRef.current = e.clientX;

            if (contentRef.current) {
                startScrollLeftRef.current = contentRef.current.scrollLeft;
            }
        },
        []
    );

    const handleMouseMove = useCallback(
        (e: MouseEvent) => {
            // istanbul ignore next
            if (!isDragging || !contentRef.current || !trackRef.current) return;

            const deltaX = e.clientX - startClientXRef.current;
            const { scrollWidth, clientWidth } = contentRef.current;
            const trackWidth = trackRef.current.clientWidth;
            const scrollRatio = scrollWidth / clientWidth;

            const newScrollLeft =
                startScrollLeftRef.current +
                (deltaX * scrollRatio) / trackWidth;
            contentRef.current.scrollLeft = newScrollLeft;
        },
        [isDragging]
    );

    const handleMouseUp = useCallback(() => {
        if (isDragging) setIsDragging(false);
    }, [isDragging]);

    useEffect(() => {
        const contentElement = contentRef.current;
        const trackElement = trackRef.current;
        // istanbul ignore next
        if (!contentElement || !trackElement) return;

        let resizeFrame: number;

        // istanbul ignore next
        const onResize = () => {
            if (resizeFrame) cancelAnimationFrame(resizeFrame);
            resizeFrame = requestAnimationFrame(() => handleResize());
        };

        const resizeObserver = new ResizeObserver(onResize);

        resizeObserver.observe(contentElement);
        resizeObserver.observe(trackElement);

        requestAnimationFrame(() => {
            handleResize();
        });

        return () => {
            resizeObserver.disconnect();
            // istanbul ignore next
            if (resizeFrame) cancelAnimationFrame(resizeFrame);
        };
    }, [handleResize]);

    useEffect(() => {
        const onWindowResize = () => handleResize();
        window.addEventListener("resize", onWindowResize);

        return () => {
            window.removeEventListener("resize", onWindowResize);
        };
    }, [handleResize]);

    useEffect(() => {
        if (isDragging) {
            document.addEventListener("mousemove", handleMouseMove);
            document.addEventListener("mouseup", handleMouseUp);
        } else {
            document.removeEventListener("mousemove", handleMouseMove);
            document.removeEventListener("mouseup", handleMouseUp);
        }
        return () => {
            document.removeEventListener("mousemove", handleMouseMove);
            document.removeEventListener("mouseup", handleMouseUp);
        };
    }, [isDragging, handleMouseMove, handleMouseUp]);

    useEffect(() => {
        const contentElement = contentRef.current;

        if (contentElement) {
            const onScroll = () => {
                handleThumbPosition();
            };

            handleThumbPosition();
            contentElement.addEventListener("scroll", onScroll);

            return () => {
                contentElement.removeEventListener("scroll", onScroll);
            };
        }
    }, [handleThumbPosition]);

    return (
        <div
            className={styles.scrollableContainer}
            data-testid="scrollContainer"
        >
            <div
                ref={contentRef}
                className={styles.contentContainer}
                data-testid="contentContainer"
            >
                {children}
            </div>
            <div
                ref={trackRef}
                className={`${styles.scrollbar} ${
                    isScrollbarVisible ? "" : styles.hiddenScrollbar
                }`}
                onClick={handleTrackClick}
                data-testid="scrollTrack"
            >
                <div
                    ref={thumbRef}
                    data-testid="scrollThumb"
                    className={styles.thumb}
                    style={{ width: `${thumbWidthRef.current}%` }}
                    onMouseDown={handleThumbMouseDown}
                />
            </div>
        </div>
    );
};

export default Scrollbar;
