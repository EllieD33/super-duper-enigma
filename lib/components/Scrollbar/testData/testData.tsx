import React from "react";

const item = (
    <div
        style={{
            minWidth: "200px",
            height: "200px",
            backgroundColor: "teal",
            margin: "5px",
        }}
    ></div>
);

export const mockContent = (
    <div style={{ display: "flex" }}>
        {Array(16)
            .fill(null)
            .map((_, index) => (
                <div key={index} style={{ margin: "0 5px" }}>
                    Item {index + 1}
                </div>
            ))}
    </div>
);
