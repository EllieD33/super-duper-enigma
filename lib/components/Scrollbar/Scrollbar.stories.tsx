import React from "react";
import type { Meta, StoryObj } from "@storybook/react";
import Scrollbar from "./Scrollbar";
import { mockContent as children } from "./testData/testData";

const meta = {
    title: "Library/Scrollbar/Scrollbar",
    component: Scrollbar,
    parameters: {},
    tags: ["autodocs"],
    argTypes: {},
} satisfies Meta<typeof Scrollbar>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
    args: {
        children,
    },
    render: (args) => <Scrollbar>{args.children}</Scrollbar>,
};
