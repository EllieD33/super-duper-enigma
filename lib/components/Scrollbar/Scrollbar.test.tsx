import React from "react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import Scrollbar from "./Scrollbar";
import { mockContent as children } from "./testData/testData";

describe("Scrollbar", () => {
    it("should render the component", () => {
        const { container } = render(<Scrollbar>{children}</Scrollbar>);
        const scrollContainer = screen.getByTestId("scrollContainer");
        expect(container).toMatchSnapshot();
        expect(scrollContainer).toBeInTheDocument();
    });

    it("should calculate the thumb width based on the container size", () => {
        render(<Scrollbar>{children}</Scrollbar>);
        const thumb = screen.getByTestId("scrollThumb");
        const thumbWidth = parseInt(thumb.style.width, 10);
        expect(thumbWidth).toBeGreaterThan(0);
    });

    it("should move thumb correctly while scrolling content", () => {
        render(<Scrollbar>{children}</Scrollbar>);
        const contentContainer = screen.getByTestId("contentContainer");
        const thumb = screen.getByTestId("scrollThumb");

        fireEvent.scroll(contentContainer, { target: { scrollLeft: 200 } });

        waitFor(() => {
            const thumbLeft = parseFloat(
                thumb.style.transform
                    .replace("translateX(", "")
                    .replace("%)", "")
            );
            expect(thumbLeft).toBeGreaterThan(0);
        });
    });
});
