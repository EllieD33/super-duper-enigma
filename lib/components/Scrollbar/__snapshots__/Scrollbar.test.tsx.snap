// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scrollbar should render the component 1`] = `
<div>
  <div
    class="scrollableContainer"
    data-testid="scrollContainer"
  >
    <div
      class="contentContainer"
      data-testid="contentContainer"
    >
      <div
        style="display: flex;"
      >
        <div
          style="margin: 0px 5px;"
        >
          Item 
          1
        </div>
        <div
          style="margin: 0px 5px;"
        >
          Item 
          2
        </div>
        <div
          style="margin: 0px 5px;"
        >
          Item 
          3
        </div>
        <div
          style="margin: 0px 5px;"
        >
          Item 
          4
        </div>
        <div
          style="margin: 0px 5px;"
        >
          Item 
          5
        </div>
        <div
          style="margin: 0px 5px;"
        >
          Item 
          6
        </div>
        <div
          style="margin: 0px 5px;"
        >
          Item 
          7
        </div>
        <div
          style="margin: 0px 5px;"
        >
          Item 
          8
        </div>
        <div
          style="margin: 0px 5px;"
        >
          Item 
          9
        </div>
        <div
          style="margin: 0px 5px;"
        >
          Item 
          10
        </div>
        <div
          style="margin: 0px 5px;"
        >
          Item 
          11
        </div>
        <div
          style="margin: 0px 5px;"
        >
          Item 
          12
        </div>
        <div
          style="margin: 0px 5px;"
        >
          Item 
          13
        </div>
        <div
          style="margin: 0px 5px;"
        >
          Item 
          14
        </div>
        <div
          style="margin: 0px 5px;"
        >
          Item 
          15
        </div>
        <div
          style="margin: 0px 5px;"
        >
          Item 
          16
        </div>
      </div>
    </div>
    <div
      class="scrollbar hiddenScrollbar"
      data-testid="scrollTrack"
    >
      <div
        class="thumb"
        data-testid="scrollThumb"
        style="width: 20%; transform: translateX(NaNpx);"
      />
    </div>
  </div>
</div>
`;
